Automate settings UI migration on plugin update (5544)

Narek13 opened the pull request:
## Issue Description

Automates migration from legacy to new settings UI on plugin update. Removes manual "Switch to New Settings" button and all related code.

## PR Description

### What Changed

**Removed:**
- `SwitchSettingsUiEndpoint` class and AJAX handler
- `should_use_the_old_ui()` method and all conditional UI branching
- Manual switch button, admin notice, and JavaScript assets
- Feature flags and environment variable checks
- Legacy database options cleanup

**Enhanced:**
- `MigrationManager` now handles automatic migration on plugin update
- Added idempotent check via `OPTION_NAME_MIGRATION_IS_DONE` flag
- Added error logging (non-blocking)
- Integrated `OnboardingProfile` updates into migration flow

### How It Works

Migration auto-triggers on `woocommerce_paypal_payments_gateway_migrate` hook:

1. Check if already migrated (skip if yes)
2. Clean up legacy UI options (`woocommerce_ppcp-settings-should-use-old-ui`, `woocommerce-ppcp-is-new-merchant`)
3. Mark onboarding complete, sync gateways
4. Run all migration classes
5. Set completion flag

### Testing

**Scenarios Verified:**
- ✅ Fresh install: No migration, no errors
- ✅ Existing merchant: Auto-migration on update
- ✅ Already migrated: Skipped (idempotent)
- ✅ Migration failure: Logged, non-blocking

**Local Testing:**
```bash
ddev snapshot              # Backup before test
# Test migration scenarios
ddev snapshot restore      # Restore and re-test
```

Narek13 committed:
refactor: remove legacy settings UI switch functionality

Remove the conditional logic that allowed merchants to manually switch
between old and new settings interfaces. This includes:

- Switch button UI in settings page header
- Admin notice promoting the new settings UI
- JavaScript asset registration for switch functionality
- AJAX endpoint handler for processing switch requests

The old UI code path and early return have been removed as the new
settings UI is now the default for all merchants.

Related to migration towards new settings UI as default.

Narek13 committed:
refactor: remove old UI detection logic across all modules

Remove the `should_use_the_old_ui()`` method and all references to it
throughout the codebase. The new settings UI is now the default and
only interface for all merchants.

Changes include:
- Removed `should_use_the_old_ui()` static method from SettingsModule
- Removed feature flag filter `'woocommerce.feature-flags.woocommerce_paypal_payments.settings_enabled'`
- Removed filter `'woocommerce_paypal_payments_should_use_the_old_ui'`
- Removed environment variable check for `PCP_SETTINGS_ENABLED`
- Removed new merchant flag check `(woocommerce-ppcp-is-new-merchant option)`
- Removed all conditional logic in other modules that branched based on UI version

Narek13 committed:
refactor: remove redundant variable

Narek13 committed:
refactor: remove manual settings UI switch endpoint

Remove the SwitchSettingsUiEndpoint class as manual UI switching is no
longer needed. Migration will be handled automatically on plugin update
in a subsequent change.

Narek13 committed:
refactor: automate settings migration on plugin update

Consolidate migration logic into `MigrationManager` and trigger it
automatically during plugin updates instead of requiring manual action.

Changes:
- Moved legacy option cleanup into `MigrationManager.migrate()`
- Added onboarding profile completion to migration flow
- Added auto-migration hook on plugin update
- Implemented idempotent check to prevent duplicate migrations
- Added warning logging for migration failures

Migration now runs automatically when merchants update the plugin,
cleaning up legacy UI options and ensuring all settings are properly
migrated to the new interface format.

Narek13 committed:
chore: run migration only during the updates

Narek13 committed:
chore: remove the switchSettingsUi from webpack config

Dinamiko commented on modules/ppcp-settings/src/Service/Migration/MigrationManager.php:
```diff
@@ -9,34 +9,80 @@
 	protected StylingSettingsMigration $styling_settings_migration;
 	protected PaymentSettingsMigration $payment_settings_migration;
+	protected OnboardingProfile $onboarding_profile;
+	protected bool $is_onboarded;
```
`$is_onboarded` seems not used in the class.

Dinamiko reviewed (approved):
PR looks good, I've approved, I just leaved an small comment.

Perhaps I'm missing an integration test that ensures the migration logic triggers and works correctly, but not sure how complex may it be to create such test so just a suggestion is case it's feasible.

Narek13 committed:
test: add integration tests for automatic settings migration

Add comprehensive test coverage for the MigrationManager to validate
automatic migration behavior during plugin updates.

Test coverage:
- API failure handling with graceful degradation
- Idempotent migration behavior (run-once guarantee)
- Migration hook skip logic when already completed
- Legacy options cleanup (woocommerce_ppcp-settings-should-use-old-ui,
  woocommerce-ppcp-is-new-merchant)
- Onboarding profile updates during migration

Tests validate that migration completes successfully in environments
without valid API credentials by leveraging early-return logic in
individual migration classes. The tests focus on core migration
functionality (cleanup, profile updates, completion flags) rather
than API-dependent data migrations.

All tests use the integration test framework with proper setup/teardown
to ensure isolated test execution and clean state between runs.

Narek13 committed:
chore: remove redundant property

puntope commented on modules/ppcp-applepay/src/ApplepayModule.php:
```diff
@@ -97,11 +96,6 @@ static function () use ( $c, $module ) {
 				$module->load_admin_assets( $c, $apple_payment_method );
 				$module->load_block_editor_assets( $c, $apple_payment_method );
-
-				if ( SettingsModule::should_use_the_old_ui() && ! $apple_payment_method->is_enabled() ) {
```
I understand why we are removing SettingsModule::should_use_the_old_ui()

But why are we removing `$apple_payment_method->is_enabled()` check?

puntope commented on modules/ppcp-compat/src/CompatModule.php:
```diff
@@ -122,12 +120,8 @@ static function () use ( $c ) {
-
 				// Store Country is set as Canada.
-				if ( $c->get( 'api.shop.country' ) !== 'CA' || ! $stay_updated ) {
+				if ( $c->get( 'api.shop.country' ) !== 'CA' || ! $settings_model->get_stay_updated() ) {
```
We can remove also the docs 

```
line 100 (checked in either old or new UI)

...

 line 102 * The "Stay updated" setting is retrieved differently based on the UI version:
		 * - Legacy UI: Retrieved from wcgateway.settings
		 * - New UI: Retrieved from settings.data.settings model
		 
		 
...
Line 116 (checked in either old or new UI).

```

puntope commented on modules/ppcp-googlepay/src/GooglepayModule.php:
```diff
@@ -151,10 +149,6 @@ static function () use ( $c, $button ) {
 				add_action(
 					'woocommerce_blocks_payment_method_type_registration',
 					function ( PaymentMethodRegistry $payment_method_registry ) use ( $c, $button ): void {
-						if ( SettingsModule::should_use_the_old_ui() && ! $button->is_enabled() ) {
```
Same as my previous comment. Are we removing also `$button->is_enabled()` how this is related to the old UI and migration?

puntope commented on modules/ppcp-settings/src/Data/Definition/TodosDefinition.php:
```diff
@@ -288,7 +288,7 @@ public function get(): array {
 	 */
 	protected function is_settings_migration_done(): bool {
-		return '1' === get_option( SwitchSettingsUiEndpoint::OPTION_NAME_MIGRATION_IS_DONE );
+		return '1' === get_option( MigrationManager::OPTION_NAME_MIGRATION_IS_DONE );
```
Maybe we can refactor this Yoda

Narek13 replied:
Kept for consistency with existing codebase style. Actually we have both :)

puntope replied:
Ok. Not a blocker. I mentioned because in modules/ppcp-settings/src/SettingsModule.php

Its using `get_option( MigrationManager::OPTION_NAME_MIGRATION_IS_DONE ) === '1'`

Narek13 committed:
chore: remove redundant comments

puntope commented on modules/ppcp-wc-gateway/services.php:
```diff
@@ -1966,18 +1965,7 @@ static function ( ContainerInterface $container ): string {
-		}
-
-		return $settings->has( 'fraudnet_enabled' ) && $settings->get( 'fraudnet_enabled' );
+		return true;
```
If this is always `true`. do we need this service ?

Narek13 replied:
This is not the scope of this PR I think, we can remove it during legacy code removal phase. I have tried to touch as few files/code as possible

puntope replied:
I see you went over FraudNet, that's why I ask. 

But agreed to change it in further PR's if needed. I still think this kind of thing always gets forgotten and is a minor change, tho.

puntope reviewed (approved):
LGTM Approved as there is no blocking changes needed. 

Left some 💅  minor comments.

Dinamiko commented on tests/integration/PHPUnit/Settings/Service/Migration/MigrationManagerTest.php:
```diff
@@ -0,0 +1,162 @@
+		update_option('woocommerce-ppcp-is-new-merchant', '0');
+
+		// Assert pre-conditions
+		$this->assertSame('yes', get_option('woocommerce_ppcp-settings-should-use-old-ui'));
+		$this->assertSame('0', get_option('woocommerce-ppcp-is-new-merchant'));
```
No need to assert these two, this is already tested by WordPress.

Narek13 replied:
If I remove these. this one fails for some reason
`$this->assertNotSame('1', get_option(MigrationManager::OPTION_NAME_MIGRATION_IS_DONE));`

Dinamiko commented on tests/integration/PHPUnit/Settings/Service/Migration/MigrationManagerTest.php:
```diff
@@ -0,0 +1,162 @@
+
+		// Verify they exist
+		$this->assertSame('yes', get_option('woocommerce_ppcp-settings-should-use-old-ui'));
+		$this->assertSame('1', get_option('woocommerce-ppcp-is-new-merchant'));
```
No need to assert these two, this is already tested by WordPress.

Dinamiko commented on tests/integration/PHPUnit/Settings/Service/Migration/MigrationManagerTest.php:
```diff
@@ -0,0 +1,162 @@
+		$this->assertSame('1', get_option(MigrationManager::OPTION_NAME_MIGRATION_IS_DONE));
+
+		// Note: Since migration check happens in the hook (not in migrate() method),
+		// the migrate() method will still run. This test needs adjustment.
```
This test seems not much useful, currently it's only asserting that `MigrationManager::OPTION_NAME_MIGRATION_IS_DONE` option is set to `1` which is done in the first line of the test by WordPress `update_option`.

I suggest to either make it test something that give us value or delete it.

Dinamiko commented on tests/integration/PHPUnit/Settings/Service/Migration/MigrationManagerTest.php:
```diff
@@ -0,0 +1,162 @@
+
+		// Assert: Legacy options NOT cleaned up (migration was skipped)
+		$this->assertSame('yes', get_option('woocommerce_ppcp-settings-should-use-old-ui'));
+		$this->assertSame('1', get_option('woocommerce-ppcp-is-new-merchant'));
```
Not sure about this test, it will also pass if you comment `do_action('woocommerce_paypal_payments_gateway_migrate', '3.2.0');`, what is the point of asserting these two options? they are already updated by WordPress via `update_option` in lines 105 and 106.

Dinamiko reviewed (changes requested):
I suggest to either remove tests `test_migration_is_idempotent` and `test_migration_hook_skips_if_already_completed` or find a way to assert a change in the state triggered by the SUT of the test.

Narek13 committed:
test: fix tests

Narek13 committed:
test: fix tests

Narek13 committed:
Merge branch 'develop' into PCP-5544-auto-trigger-migration-on-plugin-update

# Conflicts:
#	modules/ppcp-settings/services.php
#	modules/ppcp-settings/src/SettingsModule.php
#	modules/ppcp-settings/webpack.config.js
#	modules/ppcp-uninstall/services.php
#	modules/ppcp-wc-gateway/services.php

Narek13 committed:
Merge remote-tracking branch 'origin/PCP-5544-auto-trigger-migration-on-plugin-update' into PCP-5544-auto-trigger-migration-on-plugin-update

Narek13 committed:
chore: remove unnecessary uses

Narek13 committed:
test: fix tests

Narek13 committed:
chore: Remove switch UI js

Narek13 committed:
Merge remote-tracking branch 'origin/dev/PCP-5413-remove-legacy-ui' into PCP-5544-auto-trigger-migration-on-plugin-update

# Conflicts:
#	modules/ppcp-googlepay/src/GooglepayModule.php
#	modules/ppcp-settings/services.php
#	modules/ppcp-settings/src/Service/Migration/MigrationManager.php
#	modules/ppcp-status-report/src/StatusReportModule.php

Narek13 committed:
chore: Fix merge conflict

Narek13 committed:
chore: Fix wrong service call
